libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c

```diff
@@ -294,6 +318,82 @@ AwsIotProvisioningError_t _AwsIotProvisioning_SerializeRegisterThingRequestPaylo
+                                                                                               size_t * pBufferSize )
+{
+    IOT_FUNCTION_ENTRY( AwsIotProvisioningError_t, AWS_IOT_PROVISIONING_SUCCESS );
+    IotSerializerEncoderObject_t outermostPayloadEncoder = IOT_SERIALIZER_ENCODER_CONTAINER_INITIALIZER_STREAM;
+    *pBufferSize = 0;
+
+    status = _serializeCreateKeysAndCertificateRequestPayload( &outermostPayloadEncoder, NULL, 0 );
+
+    if( status != AWS_IOT_PROVISIONING_SUCCESS )
+    {
+        IOT_GOTO_CLEANUP();
+    }
+
+    /* Get the calculated required size. */
+    *pBufferSize = _pAwsIotProvisioningEncoder->getExtraBufferSizeNeeded( &outermostPayloadEncoder );
+    AwsIotProvisioning_Assert( *pBufferSize != 0 );
+
+    /* Clean the encoder object handle. */
+    _pAwsIotProvisioningEncoder->destroy( &outermostPayloadEncoder );
+
+    /* Allocate memory for the request payload based on the size required from the dry-run of serialization */
+    *pSerializationBuffer = AwsIotProvisioning_MallocPayload( *pBufferSize * sizeof( uint8_t ) );
+
+    status = _serializeCreateKeysAndCertificateRequestPayload( &outermostPayloadEncoder,
+                                                               *pSerializationBuffer,
+                                                               *pBufferSize );
+
+    IOT_FUNCTION_CLEANUP_BEGIN();
+
+    _pAwsIotProvisioningEncoder->destroy( &outermostPayloadEncoder );
+
+    IOT_FUNCTION_CLEANUP_END();
+}
+
+/*------------------------------------------------------------------*/
+
+AwsIotProvisioningError_t _AwsIotProvisioning_SerializeRegisterThingRequestPayload( const AwsIotProvisioningRegisterThingRequestInfo_t * pRequestData,
+                                                                                    uint8_t ** pSerializationBuffer,
+                                                                                    size_t * pBufferSize )
+{
+    IOT_FUNCTION_ENTRY( AwsIotProvisioningError_t, AWS_IOT_PROVISIONING_SUCCESS );
+    IotSerializerEncoderObject_t outermostPayloadEncoder = IOT_SERIALIZER_ENCODER_CONTAINER_INITIALIZER_STREAM;
+    *pBufferSize = 0;
+
+    status = _serializeRegisterThingRequestPayload( pRequestData, &outermostPayloadEncoder, NULL, 0 );
+
+    if( status != AWS_IOT_PROVISIONING_SUCCESS )
+    {
+        IOT_GOTO_CLEANUP();
+    }
+
+    /* Get the calculated required size. */
+    *pBufferSize = _pAwsIotProvisioningEncoder->getExtraBufferSizeNeeded( &outermostPayloadEncoder );
+    AwsIotProvisioning_Assert( *pBufferSize != 0 );
+
+    /* Clean the encoder object handle. */
+    _pAwsIotProvisioningEncoder->destroy( &outermostPayloadEncoder );
+
+    /* Allocate memory for the request payload based on the size required from the dry-run of serialization */
+    *pSerializationBuffer = AwsIotProvisioning_MallocPayload( *pBufferSize * sizeof( uint8_t ) );
```

Comment:
Curious when this memory gets freed and how we ensure that it never leaks? I prefer the pattern where the one allocating the memory is responsible for also freeing it. There are many patterns you can use to ensure you alloc and free symmetrically. What I do not like is when the library allocates and the application frees, this is not symmetrical and also btw. makes it impossible for the application to use the stack or statically allocated memory for this. If you let the app allocate and free then it is easy to do just that.

Reply:
I understand that concern, and yes, having this asymmetry of memory allocation is not ideal. 
The responsibility of memory allocation and release was being handled by the caller initially, but that was inter-twined with the extraneous logic of calling the serialize function **twice**, **one** for a dry serialization run with the encoder which would be used to calculate the amount of buffer required for storing the serialized data, and **the other** for performing the _actual_ serialization. 
The above details are related to the functioning (or limitation) of the serializer which, I think, does not need to be exposed in the API function implementation. Thus, I decided to encapsulate this complexity in the serializer module so that the API implementation only contains a single call for serialization to improve readability and complexity score for the API function as the goal of this PR. 

I had discussed this with @gkwicker yesterday to make sure that the memory management asymmetry is a pain we are fine with taking (_for-now_) to reduce the complexity score (or improve readability) of the API function...